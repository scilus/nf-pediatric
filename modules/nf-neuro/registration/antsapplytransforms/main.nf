process REGISTRATION_ANTSAPPLYTRANSFORMS {
    tag "$meta.id"
    label 'process_low'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        "https://scil.usherbrooke.ca/containers/scilus_latest.sif":
        "scilus/scilus:latest"}"

    input:
    tuple val(meta), path(image), path(reference), path(warp), path(affine)

    output:
    tuple val(meta), path("*__warped.nii.gz")                           , emit: warped_image
    tuple val(meta), path("*_registration_antsapplytransforms_mqc.gif") , emit: mqc, optional: true
    path "versions.yml"                                                 , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def suffix = task.ext.first_suffix ? "${task.ext.first_suffix}__warped" : "__warped"
    def suffix_qc = task.ext.suffix_qc ? "${task.ext.suffix_qc}" : ""

    def output_dtype = task.ext.output_dtype ? "-u " + task.ext.output_dtype : ""
    def dimensionality = task.ext.dimensionality ? "-d " + task.ext.dimensionality : "-d 3"
    def image_type = task.ext.image_type ? "-e " + task.ext.image_type : "-e 0"
    def interpolation = task.ext.interpolation ? "-n " + task.ext.interpolation : ""
    def default_val = task.ext.default_val ? "-f " + task.ext.default_val : ""
    def run_qc = task.ext.run_qc ? task.ext.run_qc : false

    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=$task.cpus
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1

    for image in $image;
        do \
        ext=\${image#*.}
        bname=\$(basename \${image} .\${ext})

        antsApplyTransforms $dimensionality\
                            -i \$image\
                            -r $reference\
                            -o ${prefix}__\${bname}${suffix}.nii.gz\
                            $interpolation\
                            -t $warp $affine\
                            $output_dtype\
                            $image_type\
                            $default_val

        ### ** QC ** ###
        if $run_qc;
        then
            mv $reference reference.nii.gz
            extract_dim=\$(mrinfo ${prefix}__\${bname}${suffix}.nii.gz -size)
            read sagittal_dim coronal_dim axial_dim <<< "\${extract_dim}"

            # Get the middle slice
            coronal_dim=\$((\$coronal_dim / 2))
            axial_dim=\$((\$axial_dim / 2))
            sagittal_dim=\$((\$sagittal_dim / 2))

            # Set viz params.
            viz_params="--display_slice_number --display_lr --size 256 256"
            # Iterate over images.
            for image in reference \${bname}${suffix};
            do
                scil_viz_volume_screenshot.py *\${image}.nii.gz \${image}_coronal.png \
                    --slices \$coronal_dim --axis coronal \$viz_params
                scil_viz_volume_screenshot.py *\${image}.nii.gz \${image}_sagittal.png \
                    --slices \$sagittal_dim --axis sagittal \$viz_params
                scil_viz_volume_screenshot.py *\${image}.nii.gz \${image}_axial.png \
                    --slices \$axial_dim --axis axial \$viz_params
                if [ \$image != reference ];
                then
                    title="Transformed"
                else
                    title="Reference"
                fi
                convert +append \${image}_coronal*.png \${image}_axial*.png \
                    \${image}_sagittal*.png \${image}_mosaic.png
                convert -annotate +20+230 "\${title}" -fill white -pointsize 30 \
                    \${image}_mosaic.png \${image}_mosaic.png
                # Clean up.
                rm \${image}_coronal*.png \${image}_sagittal*.png \${image}_axial*.png
            done
            # Create GIF.
            convert -delay 10 -loop 0 -morph 10 \
                \${bname}${suffix}_mosaic.png reference_mosaic.png \${bname}${suffix}_mosaic.png \
                ${prefix}_\${bname}${suffix_qc}_registration_antsapplytransforms_mqc.gif
            # Clean up.
            rm *_mosaic.png
        fi
    done

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        ants: \$(antsRegistration --version | grep "Version" | sed -E 's/.*v([0-9]+\\.[0-9]+\\.[0-9]+).*/\\1/')
        mrtrix: \$(mrinfo -version 2>&1 | sed -n 's/== mrinfo \\([0-9.]\\+\\).*/\\1/p')
        imagemagick: \$(magick -version | sed -n 's/.*ImageMagick \\([0-9]\\{1,\\}\\.[0-9]\\{1,\\}\\.[0-9]\\{1,\\}\\).*/\\1/p')
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def suffix = task.ext.first_suffix ? "${task.ext.first_suffix}__warped" : "__warped"

    """
    antsApplyTransforms -h

    for image in $image;
        do \
        ext=\${image#*.}
        bname=\$(basename \${image} .\${ext})

        touch ${prefix}__\${bname}${suffix}.nii.gz
    done

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        ants: \$(antsRegistration --version | grep "Version" | sed -E 's/.*v([0-9]+\\.[0-9]+\\.[0-9]+).*/\\1/')
        mrtrix: \$(mrinfo -version 2>&1 | sed -n 's/== mrinfo \\([0-9.]\\+\\).*/\\1/p')
        imagemagick: \$(magick -version | sed -n 's/.*ImageMagick \\([0-9]\\{1,\\}\\.[0-9]\\{1,\\}\\.[0-9]\\{1,\\}\\).*/\\1/p')
    END_VERSIONS
    """
}
