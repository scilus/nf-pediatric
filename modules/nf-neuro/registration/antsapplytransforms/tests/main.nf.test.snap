{
    "registration - antsapplytransforms": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__b0b0__warped.nii.gz:md5,3ae49aacab66fb2378da92155453b1be"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0b0_to_template_registration_antsapplytransforms_mqc.gif:md5,2e97314c320a27e4271c960c50895dfb"
                    ]
                ],
                "2": [
                    "versions.yml:md5,3414a40c02551052589b50e4eb35a3da"
                ],
                "mqc": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0b0_to_template_registration_antsapplytransforms_mqc.gif:md5,2e97314c320a27e4271c960c50895dfb"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,3414a40c02551052589b50e4eb35a3da"
                ],
                "warped_image": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test__b0b0__warped.nii.gz:md5,3ae49aacab66fb2378da92155453b1be"
                    ]
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-03-14T18:15:20.750783"
    },
    "registration - antsapplytransforms - stub-run": {
        "content": [
            [
                "versions.yml:md5,3414a40c02551052589b50e4eb35a3da"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.3"
        },
        "timestamp": "2025-02-23T16:23:22.94059003"
    }
}